File: FREQNESS_Toolbox/FREQNESS_Functions/FREQNESS_NetworkEstimation.m
function [FREQ] = FREQNESS_NetworkEstimation(data, frex, srate, varargin)

% ========================================================================

%  FREQUENCY-RESOLVED NETWORK ESTIMATION TOOLBOX
%
%  If you use this toolbox, please cite:
%  Rosso, M., Fernández‐Rubio, G., Keller, P. E., Brattico, E., Vuust, P., Kringelbach, M. L., & Bonetti, L. (2025).
%  FREQ‐NESS Reveals the Dynamic Reconfiguration of Frequency‐Resolved Brain Networks During Auditory Stimulation.
%  Advanced Science, 2413195.
%  https://doi.org/10.1002/advs.202413195
%
% ========================================================================
%
%  This script takes a multivariate dataset (channels × time matrix)
%  and produces a frequency-resolved brain network landscape via
%  Generalized Eigendecomposition (GED).
%
% ------------------------------------------------------------------------
%  INPUT ARGUMENTS:
% ------------------------------------------------------------------------
%  - data   : 2D matrix in channels-by-time format, or
%             3D matrix in channels-by-time-by-subjects format
%             NOTE: While this approach can be applied to scalp sensor data,
%             source-reconstructed brain voxel data is necessary to
%             interpret the output as brain networks.
%
%  - frex   : Vector of frequencies (ascending order).
%
%  - srate  : Sampling rate (Hz).
%
%  The following parameters can be specified using name-value pairs:
%
%  - 'duration'        : Duration of the recording to analyze (seconds).
%                        Default: full data duration.
%
%  - 'fwidth'          : Full-width at half-maximum (FWHM) of the frequency filter.
%                        Default: computed based on reference values.
%
%  - 'filter'          : Filter design method. Options:
%                        'logarithmic' (default) or 'linear'.
%
%  - 'regularisation'  : Shrinkage factor for covariance matrix regularization.
%                        Default: 0.01.
%
%  - 'ncomps'          : Number of components to retain per frequency.
%                        Default: 30.
%
%  - 'bad_segments'    : Vector of sample indices (e.g., 1-based timepoints)
%                        to be excluded from the analysis.
%                        These will be removed from covariance computations.
%
%  - 'rescale'         : Logical flag controlling automatic rescaling of
%                        very low-amplitude data. Default: false.
%
% ------------------------------------------------------------------------
%  OUTPUT ARGUMENTS:
% ------------------------------------------------------------------------
%  - FREQ       : Structure containing eigenvalues, eigenvectors, spatial
%                 activation patterns, network time series, and additional 
%                 variables which will be used by second-order functions in
%                 the toolbox.
%                 Info on the specific fields is found in the comments where
%                 the respective variables are assigned to GED.
%
% ------------------------------------------------------------------------
%  AUTHORS:
%  Mattia Rosso & Leonardo Bonetti
%  mattia.rosso@clin.au.dk
%  leonardo.bonetti@clin.au.dk; leonardo.bonetti@psych.ox.ac.uk
%  Center for Music in the Brain, Aarhus University
%  Centre for Eudaimonia and Human Flourishing, Linacre College, University of Oxford
%  Aarhus (DK), Oxford (UK), 12/06/2025

% ========================================================================


% NOTE: credit for the GED implementation goes to Mike X Cohen
% (Cohen, 2022 - 'A tutorial on generalized eigendecomposition for denoising,
% contrast enhancement, and dimension reduction in multichannel electrophysiology')


%% Check mandatory inputs

if nargin < 3
    error('data, frex, and srate are required inputs.');
end

if ~isnumeric(data) || ~isreal(data) || isempty(data) || ...
        ndims(data) > 3 || any(~isfinite(data(:)))
    error(['data must be a non-empty, real, finite numeric array in ' ...
        '[channels x time] or [channels x time x participants] format.']);
end

if ~isnumeric(srate) || ~isreal(srate) || ~isscalar(srate) || ...
        ~isfinite(srate) || srate <= 0
    error('srate must be one positive finite scalar.');
end

if ~isnumeric(frex) || ~isreal(frex) || ~isvector(frex) || isempty(frex) || ...
        any(~isfinite(frex)) || any(frex <= 0)
    error('frex must be a non-empty vector of positive finite frequencies.');
end

if any(frex >= srate/2)
    error('All frequencies in frex must be lower than the Nyquist frequency.');
end

%% Handle Optional Arguments

% Define default values
opts = struct('duration', [], 'fwidth', [], 'filter', 'logarithmic', ...
    'regularisation', 0.01, 'ncomps', 30, 'bad_segments',[], ...
    'rescale',false);

% Parse name-value pair arguments
opts = parse_name_value_pairs(opts, varargin{:});

% If duration is empty, use full data duration
if isempty(opts.duration)
    opts.duration = size(data, 2) / srate;
end

% Assign values
duration       = opts.duration;
fwidth         = opts.fwidth;
filter         = opts.filter;
regularisation = opts.regularisation;
ncomps         = opts.ncomps;
idx2remove     = opts.bad_segments; % in the current version, these apply to all subjects in the input
rescale        = opts.rescale;

% Validate optional inputs
if ~isempty(duration) && (~isnumeric(duration) || ~isreal(duration) || ...
        ~isscalar(duration) || ~isfinite(duration) || duration <= 0)
    error('duration must be one positive finite scalar expressed in seconds.');
end

if ~(ischar(filter) || (isstring(filter) && isscalar(filter))) || ...
        ~(strcmpi(filter,'logarithmic') || strcmpi(filter,'linear'))
    error('filter must be either ''logarithmic'' or ''linear''.');
end

if ~isnumeric(regularisation) || ~isreal(regularisation) || ...
        ~isscalar(regularisation) || ~isfinite(regularisation) || ...
        regularisation < 0 || regularisation >= 1
    error('regularisation must be one finite scalar in the interval [0, 1).');
end

if ~isnumeric(ncomps) || ~isreal(ncomps) || ~isscalar(ncomps) || ...
        ~isfinite(ncomps) || ncomps ~= round(ncomps) || ncomps < 1
    error('ncomps must be one positive integer.');
end


if ncomps > size(data,1)
    error('ncomps cannot exceed the number of input channels or voxels.');
end

if ~islogical(rescale) || ~isscalar(rescale)
    error('rescale must be one logical value (true or false).');
end

% Compute necessary parameters
pnts2keep = duration * srate;

if abs(pnts2keep - round(pnts2keep)) > eps(max(1,pnts2keep))
    error('duration multiplied by srate must define an integer number of samples.');
end
pnts2keep = round(pnts2keep);

% Check duration
if pnts2keep > size(data,2)
    error('Requested duration exceeds the length of the data.');
end

if pnts2keep < 2
    error('The requested duration must contain at least two samples.');
end

if ~isempty(idx2remove) && (~isnumeric(idx2remove) || ~isreal(idx2remove) || ...
        ~isvector(idx2remove) || any(~isfinite(idx2remove)) || ...
        any(idx2remove ~= round(idx2remove)))
    error('bad_segments must contain finite integer sample indices.');
end

idx2remove = unique(idx2remove(:)','stable');

% Check that the range of the bad segments falls within the requested duration
if any(idx2remove < 1) || any(idx2remove > pnts2keep)
    error('One or more indices in bad_segments fall outside the analyzed time range.');
end

% Define good indexes to keep for computing the covariance matrices
idx2cov = setdiff(1:pnts2keep,idx2remove);
if numel(idx2cov) < 2
    error('Fewer than two samples remain for covariance estimation.');
end

% Re-sort frequency vector in ascending order
frex = sort(frex,'ascend');
nfrex = length(frex);

% Handle missing filter width (based on Rosso et al., 2025 - Advanced Science)
if isempty(fwidth)

    ref_frex = 2.439; % set a reference frequency
    ref_fwhm = 0.35;  % set a reference filter width
    nfrex_above = 80; % how many frequencies above the reference
    nfrex_below = 6;  % how many frequencies below the reference
    % Compute the vector of frequencies
    frex_above = linspace(ref_frex, ref_frex * (nfrex_above / 2), nfrex_above);
    frex_below = linspace(ref_frex / 2, ref_frex / (2 * nfrex_below), nfrex_below);
    frex_all = [frex_below(end:-1:1), frex_above];
    % Compute the vector of filter width (log-spaced)
    fwhm_above = logspace(log10(ref_fwhm), log10(ref_fwhm * (nfrex_above)), nfrex_above);
    fwhm_below = logspace(log10(ref_fwhm), log10(ref_fwhm / (nfrex_below)), nfrex_below + 1);
    fwidth_all = [fwhm_below(end:-1:2), fwhm_above];
    % Match user-requested frequencies with the reference frequencies from Rosso et al. (2025)
    [~, idx] = min(abs(frex_all - frex(1))); % matching the lowest user-defined frequency
    fwidth = fwidth_all(idx);

    % Compute filter widths
    if strcmpi(filter, 'logarithmic')
        fwidth_all = logspace(log10(fwidth), log10(fwidth * nfrex), nfrex);
    else
        fwidth_all = linspace(fwidth, fwidth * nfrex, nfrex);
    end

else

    % Assign input filter width vector, if given by the user
    if isscalar(fwidth)
        fwidth_all = repmat(fwidth,1,nfrex);
    elseif isvector(fwidth) && numel(fwidth)==nfrex
        fwidth_all = fwidth(:)';
    else
        error('fwidth must be a scalar or contain one value per frequency.');
    end

end

if ~isnumeric(fwidth_all) || ~isreal(fwidth_all) || ...
        any(~isfinite(fwidth_all)) || any(fwidth_all <= 0)
    error('fwidth must contain positive finite values.');
end


%% Input data check

% Check input size to verify if more than one participant is being analyzed
nsubs = size(data,3);

% Occasionally, anatomical source reconstruction can return extremely low
% values. This can cause unreliable source separation due to the
% scale of the regularization factor of the covariance matrices.

fprintf('\nFREQNESS Network Estimation: estimating frequency-resolved networks for %d participants.\n', nsubs);
scale_factors = ones(1,nsubs);
for subi = 1:nsubs

    % Extract data for this participant
    this_data = data(:,1:pnts2keep,subi);

    % Robust estimate of the scale of the data
    scale_ref = median(abs(this_data(:)));

    if scale_ref == 0
        error(['The median absolute data value is zero for participant #' ...
            num2str(subi) '. Data scale cannot be estimated reliably.']);
    end

    % Detect current order of magnitude
    order_mag = floor(log10(scale_ref));

    if order_mag < -2
        warning(['The scale of your data is very low (order of magnitude: 10e' num2str(order_mag) ').' ...
            ' This can cause unreliable network estimation.']);

        if rescale
            % Define desired target order
            scale = 10^abs(order_mag - 1);  % re-scale to bring values into the hundreds range

            % Assign re-scaled data
            data(:,1:pnts2keep,subi) = scale * this_data;
            scale_factors(subi) = scale;
        end
    end

end

%% Generalized eigendecomposition (GED) 

% Initialize output structure
FREQ = [];

% Initialize GED outputs
GEDevals = zeros(ncomps, nfrex, nsubs);
[GEDevecs, GEDpats] = deal(zeros(size(data,1), ncomps, nfrex, nsubs));
GEDts = zeros(ncomps, pnts2keep, nfrex, nsubs);

% Loop over participants (also works with one single participant)
for subi = 1:nsubs
    disp(['FREQNESS Network Estimation: participant #' num2str(subi)]);

    % Assign broadband source data
    broadData = data(:, 1:pnts2keep, subi);

    % Remove bad segments, if any bad segments is provided as input
    if ~isempty(idx2remove)
        pct_removed = 100 * length(idx2remove) / pnts2keep;
        disp(['Removing ' num2str(length(idx2remove)) ' bad timepoints ' ...
            '(' num2str(pct_removed) '% of the input data).']);
    end


    % Compute and regularise the broadband covariance matrix once
    covR = cov(broadData(:,idx2cov)');
    evalsR = eig(covR);
    covR = (1-regularisation)*covR + ...
        regularisation * mean(evalsR) * eye(size(covR));

    % Perform frequency-resolved brain network separation via GED
    for frexi = 1:nfrex % loop over input frequencies
        disp(['Estimating network at ' num2str(frex(frexi)) ' Hz'])

        % Filter data
        narrowData = filterFGx(broadData,srate,frex(frexi),fwidth_all(frexi),0); % turn the last argument to 1 to visualize the filter

        % Compute narrowband covariance matrix
        covS = cov(narrowData(:,idx2cov)');

        % Regularisation (to bring the covariance matrix to full rank)
        covS = covS  + 1e-6*eye(size(covS)); % regularise the S covariance matrix by adding a small perturbation/noise

        % Eigendecomposition
        [evecs,evals] = eig(covS ,covR);
        [evals,sidx]  = sort( diag(evals),'descend' ); % the first output returns sorted evals extracted from diagonal
        evecs = evecs(:,sidx);          % sort eigenvectors
        if ~isfinite(sum(evals)) || abs(sum(evals)) < realmin
            error('Generalized eigenvalues cannot be normalized because their sum is zero or non-finite.');
        end
        evals = evals.*100./sum(evals); % normalize eigenvalues to percent variance explained
        % Assign temporary variables to output
        GEDevecs(:,:,frexi,subi) = evecs(:,1:ncomps);
        GEDevals(:,frexi,subi) = evals(1:ncomps);

        % Brain networks' spatial activation patterns and time series
        for compi = 1:ncomps

            % Compute spatial activation patterns and flip sign
            GEDpats(:,compi,frexi,subi) = covS * GEDevecs(:,compi,frexi,subi); % get component
            [~,idxmax] = max(abs(GEDpats(:,compi,frexi,subi)));     % find max magnitude
            GEDpats(:,compi,frexi,subi)  = GEDpats(:,compi,frexi,subi) * sign(GEDpats(idxmax,compi,frexi,subi)); % possible sign flip

            % Compute the network's activation time series
            GEDts(compi,:,frexi,subi) = GEDevecs(:,compi,frexi,subi)' * broadData;

        end

    end

end

% Assign outputs to structure
FREQ.evals = GEDevals;    % eigenvalues
FREQ.evecs = GEDevecs;    % eigenvectors
FREQ.pats  = GEDpats;     % spatial activation patterns
FREQ.ts    = GEDts;       % time series
FREQ.frex  = frex;        % to make use of the analyzed frequencies later on
FREQ.fwhm  = fwidth_all;  % to make use of the corresponding filter widths
FREQ.srate = srate;       % re-assign to output, so it can be used by secondary functions
FREQ.duration = pnts2keep/srate;        % analyzed duration in seconds
FREQ.bad_segments = idx2remove;         % excluded sample indices
FREQ.regularisation = regularisation;   % covariance shrinkage factor
FREQ.scale_factors = scale_factors;     % participant-wise internal scale factors

end


%% Helper Function: Parse Name-Value Pairs
function opts = parse_name_value_pairs(opts, varargin)
if mod(length(varargin), 2) ~= 0
    error('Arguments must be given as name-value pairs.');
end
for i = 1:2:length(varargin)
    name = lower(varargin{i});
    if isfield(opts, name)
        opts.(name) = varargin{i+1};
    else
        error(['Unrecognized argument: ', name]);
    end
end
end

%%

File: FREQNESS_Toolbox/python/src/freqness/FREQNESS_NetworkEstimation.py
"""Frequency-resolved network estimation via generalized eigendecomposition.

If you use this toolbox, please cite:
Rosso, M., Fernández-Rubio, G., Keller, P. E., Brattico, E., Vuust, P.,
Kringelbach, M. L., & Bonetti, L. (2025). FREQ-NESS Reveals the Dynamic
Reconfiguration of Frequency-Resolved Brain Networks During Auditory
Stimulation. Advanced Science, 2413195.
https://doi.org/10.1002/advs.202413195

Toolbox authors
---------------
Mattia Rosso - Center for Music in the Brain, Aarhus University -
mattia.rosso@clin.au.dk
Chiara Malvaso - University of Bologna - chiara.malvaso2@unibo.it
Leonardo Bonetti - Center for Music in the Brain, Aarhus University; Centre
for Eudaimonia and Human Flourishing, Linacre College, University of Oxford -
leonardo.bonetti@clin.au.dk; leonardo.bonetti@psych.ox.ac.uk
"""

from __future__ import annotations

from dataclasses import dataclass
import warnings

import numpy as np
from numpy.typing import ArrayLike, NDArray
from scipy.linalg import eigh

from .filterFGx import filterFGx


FloatArray = NDArray[np.float64]


@dataclass(slots=True)
class FREQNESSResult:
    """Outputs of :func:`FREQNESS_NetworkEstimation`.

    Array dimensions intentionally mirror the MATLAB implementation. The final
    subject dimension is retained for single-participant inputs.
    """

    evals: FloatArray
    evecs: FloatArray
    pats: FloatArray
    ts: FloatArray
    frex: FloatArray
    fwhm: FloatArray
    srate: float
    duration: float
    bad_segments: NDArray[np.int64]
    regularisation: float
    scale_factors: FloatArray


def _as_subject_array(data: ArrayLike) -> FloatArray:
    values = np.asarray(data)
    if values.ndim == 2:
        values = values[:, :, np.newaxis]
    if values.ndim != 3 or min(values.shape) == 0:
        raise ValueError(
            "data must have shape (variables, time) or "
            "(variables, time, participants)"
        )
    if not np.issubdtype(values.dtype, np.number) or np.iscomplexobj(values):
        raise TypeError("data must contain real numeric values")
    if not np.all(np.isfinite(values)):
        raise ValueError("data contains NaN or infinite values")
    return np.asarray(values, dtype=float).copy()


def _validate_frequencies(frex: ArrayLike, srate: float) -> FloatArray:
    frequencies = np.asarray(frex, dtype=float)
    if frequencies.ndim != 1 or frequencies.size == 0:
        raise ValueError("frex must be a non-empty one-dimensional vector")
    if not np.all(np.isfinite(frequencies)) or np.any(frequencies <= 0):
        raise ValueError("frex must contain positive finite values")
    if np.any(frequencies >= srate / 2):
        raise ValueError("all frequencies must be lower than the Nyquist frequency")
    return np.sort(frequencies)


def _filter_widths(
    frequencies: FloatArray,
    fwidth: ArrayLike | None,
    filter_type: str,
) -> FloatArray:
    n_frequencies = frequencies.size
    if fwidth is not None:
        widths = np.asarray(fwidth, dtype=float)
        if widths.ndim == 0:
            widths = np.repeat(widths.item(), n_frequencies)
        elif widths.ndim == 1 and widths.size == n_frequencies:
            widths = widths.copy()
        else:
            raise ValueError(
                "fwidth must be a scalar or contain one value per frequency"
            )
        if not np.all(np.isfinite(widths)) or np.any(widths <= 0):
            raise ValueError("fwidth must contain positive finite values")
        return widths

    reference_frequency = 2.439
    reference_fwhm = 0.35
    n_above = 80
    n_below = 6
    frequencies_above = np.linspace(
        reference_frequency, reference_frequency * (n_above / 2), n_above
    )
    frequencies_below = np.linspace(
        reference_frequency / 2,
        reference_frequency / (2 * n_below),
        n_below,
    )
    reference_frequencies = np.concatenate(
        [frequencies_below[::-1], frequencies_above]
    )
    fwhm_above = np.logspace(
        np.log10(reference_fwhm),
        np.log10(reference_fwhm * n_above),
        n_above,
    )
    fwhm_below = np.logspace(
        np.log10(reference_fwhm),
        np.log10(reference_fwhm / n_below),
        n_below + 1,
    )
    reference_widths = np.concatenate([fwhm_below[:0:-1], fwhm_above])
    lowest_index = int(
        np.argmin(np.abs(reference_frequencies - frequencies[0]))
    )
    lowest_width = reference_widths[lowest_index]

    if filter_type.lower() == "logarithmic":
        return np.logspace(
            np.log10(lowest_width),
            np.log10(lowest_width * n_frequencies),
            n_frequencies,
        )
    if filter_type.lower() == "linear":
        return np.linspace(
            lowest_width,
            lowest_width * n_frequencies,
            n_frequencies,
        )
    raise ValueError("filter must be 'logarithmic' or 'linear'")


def _duration_samples(duration: float | None, n_time: int, srate: float) -> int:
    if duration is None:
        return n_time
    if not np.isscalar(duration) or not np.isfinite(duration) or duration <= 0:
        raise ValueError("duration must be a positive finite scalar in seconds")
    requested = float(duration) * srate
    rounded = int(round(requested))
    if not np.isclose(requested, rounded):
        raise ValueError("duration * srate must define an integer number of samples")
    if rounded > n_time:
        raise ValueError("requested duration exceeds the length of the data")
    if rounded < 2:
        raise ValueError("the requested duration contains fewer than two samples")
    return rounded


def _good_sample_indices(
    bad_segments: ArrayLike | None,
    n_time: int,
) -> tuple[NDArray[np.int64], NDArray[np.int64]]:
    if bad_segments is None:
        bad_one_based = np.empty(0, dtype=np.int64)
    else:
        raw = np.asarray(bad_segments)
        if raw.ndim != 1 or not np.issubdtype(raw.dtype, np.number):
            raise ValueError("bad_segments must be a vector of integer sample indices")
        if not np.all(np.isfinite(raw)) or not np.all(raw == np.round(raw)):
            raise ValueError("bad_segments must contain finite integer sample indices")
        bad_one_based = np.unique(raw.astype(np.int64))
        if np.any(bad_one_based < 1) or np.any(bad_one_based > n_time):
            raise ValueError(
                "one or more bad_segments indices fall outside the analyzed time range"
            )

    mask = np.ones(n_time, dtype=bool)
    mask[bad_one_based - 1] = False
    good = np.flatnonzero(mask)
    if good.size < 2:
        raise ValueError("fewer than two samples remain for covariance estimation")
    return good, bad_one_based


def _prepare_scale_factors(data: FloatArray, rescale: bool) -> FloatArray:
    factors = np.ones(data.shape[2], dtype=float)
    for subject in range(data.shape[2]):
        scale_reference = float(np.median(np.abs(data[:, :, subject])))
        if scale_reference == 0:
            raise ValueError(
                f"participant {subject + 1} has a median absolute data value of zero"
            )
        order = int(np.floor(np.log10(scale_reference)))
        if order < -2:
            warnings.warn(
                f"Participant {subject + 1} data have a low scale "
                f"(order of magnitude 10e{order}); covariance regularization "
                "may affect network estimation.",
                UserWarning,
                stacklevel=3,
            )
            if rescale:
                factors[subject] = 10.0 ** abs(order - 1)
                data[:, :, subject] *= factors[subject]
    return factors


def FREQNESS_NetworkEstimation(
    data: ArrayLike,
    frex: ArrayLike,
    srate: float,
    *,
    duration: float | None = None,
    fwidth: ArrayLike | None = None,
    filter: str = "logarithmic",
    regularisation: float = 0.01,
    ncomps: int = 30,
    bad_segments: ArrayLike | None = None,
    rescale: bool = False,
) -> FREQNESSResult:
    """Estimate frequency-resolved networks using generalized eigendecomposition.

    Parameters follow the MATLAB function of the same name. ``bad_segments``
    deliberately uses MATLAB-compatible one-based sample indices. Low-amplitude
    data are never rescaled interactively; pass ``rescale=True`` to apply the
    MATLAB function's automatic scale factor after its warning condition.
    """
    if not np.isscalar(srate) or not np.isfinite(srate) or srate <= 0:
        raise ValueError("srate must be a positive finite scalar")
    if (
        not np.isscalar(regularisation)
        or not np.isfinite(regularisation)
        or regularisation < 0
        or regularisation >= 1
    ):
        raise ValueError("regularisation must be in the interval [0, 1)")
    if not isinstance(ncomps, (int, np.integer)) or ncomps < 1:
        raise ValueError("ncomps must be a positive integer")
    if not isinstance(rescale, (bool, np.bool_)):
        raise TypeError("rescale must be a boolean")

    subject_data = _as_subject_array(data)
    n_variables, n_time, n_subjects = subject_data.shape
    if ncomps > n_variables:
        raise ValueError("ncomps cannot exceed the number of input variables")

    sampling_rate = float(srate)
    frequencies = _validate_frequencies(frex, sampling_rate)
    widths = _filter_widths(frequencies, fwidth, filter)
    n_samples = _duration_samples(duration, n_time, sampling_rate)
    subject_data = subject_data[:, :n_samples, :]
    good_samples, bad_one_based = _good_sample_indices(bad_segments, n_samples)
    scale_factors = _prepare_scale_factors(subject_data, bool(rescale))

    n_frequencies = frequencies.size
    eigenvalues = np.zeros((ncomps, n_frequencies, n_subjects), dtype=float)
    eigenvectors = np.zeros(
        (n_variables, ncomps, n_frequencies, n_subjects), dtype=float
    )
    patterns = np.zeros_like(eigenvectors)
    time_series = np.zeros(
        (ncomps, n_samples, n_frequencies, n_subjects), dtype=float
    )

    for subject in range(n_subjects):
        broadband = subject_data[:, :, subject]
        covariance_r = np.atleast_2d(
            np.cov(broadband[:, good_samples], rowvar=True, ddof=1)
        )
        covariance_r_eigenvalues = np.linalg.eigvalsh(covariance_r)
        covariance_r = (
            (1.0 - regularisation) * covariance_r
            + regularisation
            * np.mean(covariance_r_eigenvalues)
            * np.eye(n_variables)
        )

        for frequency_index, (frequency, width) in enumerate(
            zip(frequencies, widths, strict=True)
        ):
            narrowband, _, _ = filterFGx(
                broadband,
                sampling_rate,
                float(frequency),
                float(width),
            )
            covariance_s = np.atleast_2d(
                np.cov(narrowband[:, good_samples], rowvar=True, ddof=1)
            )
            covariance_s = covariance_s + 1e-6 * np.eye(n_variables)

            all_values, all_vectors = eigh(
                covariance_s,
                covariance_r,
                check_finite=False,
            )
            order = np.argsort(all_values)[::-1]
            all_values = np.real(all_values[order])
            all_vectors = np.real(all_vectors[:, order])
            value_sum = np.sum(all_values)
            if np.isclose(value_sum, 0.0):
                raise np.linalg.LinAlgError(
                    "generalized eigenvalues sum to zero and cannot be normalized"
                )
            all_values = all_values * 100.0 / value_sum

            retained_vectors = all_vectors[:, :ncomps]
            eigenvalues[:, frequency_index, subject] = all_values[:ncomps]
            eigenvectors[:, :, frequency_index, subject] = retained_vectors
            time_series[:, :, frequency_index, subject] = (
                retained_vectors.T @ broadband
            )

            retained_patterns = covariance_s @ retained_vectors
            for component in range(ncomps):
                pattern = retained_patterns[:, component]
                largest = int(np.argmax(np.abs(pattern)))
                sign = np.sign(pattern[largest])
                if sign != 0:
                    pattern = pattern * sign
                patterns[:, component, frequency_index, subject] = pattern

    return FREQNESSResult(
        evals=eigenvalues,
        evecs=eigenvectors,
        pats=patterns,
        ts=time_series,
        frex=frequencies,
        fwhm=widths,
        srate=sampling_rate,
        duration=n_samples / sampling_rate,
        bad_segments=bad_one_based,
        regularisation=float(regularisation),
        scale_factors=scale_factors,
    )


__all__ = ["FREQNESS_NetworkEstimation", "FREQNESSResult"]
